Observer/SalesQuoteProductAddAfter.php

```diff
@@ -0,0 +1,161 @@
+            'AddedItemPrice' => (float) $addedProduct->get_price(),
+            'AddedItemQuantity' => (int) $addedItem->getQty(),
+            'AddedItemProductID' => (int) $addedProduct->getId(),
+            'AddedItemProductName' => (string) $addedItem->getName(),
+            'AddedItemSku' => (string) $addedProduct->getSku(),
+            'AddedItemUrl' => (string) $addedProduct->getProductUrl()
+        );
+
+        $klAddedToCartPayload = array_merge(
+            $this->klBuildCartData( $quote, $addedItem ),
+            $addedItemData
+        );
+
+        // Creating a custom session variable in the checkout session since quoteId is not set at this point for guest checkouts
+        $this->getCheckoutSession()->setKlAddedToCartKey( $klAddedToCartPayload );
+    }
+
+    /**
+     * Helper function to add items from cart to Added to Cart payload
+     * @param $quote
+     * @param $addedItem
+     * @return array
+     */
+    public function klBuildCartData($quote, $addedItem)
+    {
+        $cartItems = $quote->getItems() ?? array();
+        $cartQty = 0;
+        $items = array();
+        $cartItemNames = array();
+        $cartItemCategories = array();
+
+        foreach($cartItems as $item) {
+            $product = $item->getProduct();
+            $cartItemId = $product->getId();
+            $itemCategories = $this->getCategoryName($product->getCategoryIds());
+            $itemName = $item->getName();
+            $currentProduct = array(
+                'Categories' => (array) $itemCategories,
+                'ImageUrlKey' => (string) $product->getData('small_image'),
+                'ProductId' => (int) $cartItemId,
+                'Price' => (float) $product->getPrice(),
+                'Title' => (string) $itemName,
+                'Description' => (string) strip_tags($product->getDescription()),
+                'Url' => (string) $product->getProductUrl(),
+                'Quantity' => (int) $item->getQty()
+            );
+            $cartQty += $item->getQty();
+            array_push($items, $currentProduct);
+            array_push($cartItemNames, $itemName);
+            $cartItemCategories = $this->uniqueArrayOfStrings($cartItemCategories, $itemCategories);
+        }
+
+        return array(
+            '$value' => (float) $quote->getBaseGrandTotal() + $addedItem->getPrice(),
+            'ItemNames' => (array) $cartItemNames,
+            'Items' => (array) $items,
+            'ItemCount' => (int) $cartQty,
+            'Categories' => (array) $cartItemCategories
+        );
+    }
+
+    /**
+     * Retrieves category names from category IDs
+     * @param array $categoryIds
+     * @return array
+     */
+    public function getCategoryName(array $categoryIds)
+    {
+        $categoryFactory = $this->_categoryFactory->create();
+        $categoryNames = [];
+        foreach ( $categoryIds as $id ) {
+            $category = $categoryFactory->load($id);
```

Comment:
Load in the loop isn't good idea at all. It will significantly reduce performance of adding to cart in case if product assigned to let's say 100 categories

Reply:
@ihor-sviziev I'm moving the hydration categoryNames to the cron so as to ensure heavy lifting is done later without affecting end user experience